### DISCUSSION TITLE
Schaff Trend Cycle

### DISCUSSION THREAD

**Discussion by myalgomate** [DaveSkender/Stock.Indicators#570] [Indicators]
The Schaff Trend Cycle (STC) is a charting indicator that is commonly used to identify market trends and provide buy and sell signals to traders. Developed in 1999 by noted currency trader Doug Schaff, STC is a type of oscillator and is based on the assumption that, regardless of time frame, currency trends accelerate and decelerate in cyclical patterns.

Some reference to check details on indicators:

https://patternswizard.com/schaff-trend-cycle-indicator/
https://www.investopedia.com/articles/forex/10/schaff-trend-cycle-indicator.asp
https://www.tradingpedia.com/forex-trading-indicators/schaff-trend-cycle

**Comment by DaveSkender** [comment]
Thanks for the recommendation and info.  Tracking dev here: #571

**Reply by DaveSkender** [reply]
This was implemented in version 1.18.0

**Comment by DaveSkender** [comment]
On reading about this a bit more, STC is just a [Slow Stochastic](https://daveskender.github.io/Stock.Indicators/indicators/Stoch/#content) of a [MACD](https://daveskender.github.io/Stock.Indicators/indicators/Macd/#content).  Given that these two are already in the library, you can compute things like this on your own; for example, STC is:

```csharp
IEnumerable<StochResult> stochMacd = quotes
  .GetMacd(fastPeriods, slowPeriods, 1)
  .Where(x => x.Macd != null)
  .Select(x => new Quote    // create synthetic Quote of MACD
  {
      Date = x.Date,
      High = (decimal)x.Macd,
      Low = (decimal)x.Macd,
      Close = (decimal)x.Macd
  })
  .GetStoch(cyclePeriods, 1, 3);
```

**Reply by DaveSkender** [reply]
With that said, when you do it this way the results will not be the same length as the original `quotes` due to the interim pruning needed for the Stoch calculation.  As such, I'm still building this as a standalone indicator.

**Reply by DaveSkender** [reply]
I provide this example just to demonstrate how you can generally create your own composite [indicator of indicators](https://daveskender.github.io/Stock.Indicators/guide/#generating-indicator-of-indicators).

**Comment by turok2000** [comment]
Прошу подсказать как мне встроить в ваш код такую конструкцию?
Can you please tell me how to build this construct into your code?
```
function SchaffTC(period, s_len, l_len)
{
	/// Schaff Trend Cycle
	/// @link https://forum.amibroker.com/t/schaff-trend-indicator-stc/13355/6
	local MCD, frac1, Frac2, PF, PFF;
	MCD = MACD(s_len, l_len);

	//StochK of MACD
	C = H = L = MCD;
	frac1 = StochK(period, 1);

	//Smoothed calculation
	PF = AMA(frac1, factor = 0.5);

	//StochK of PF
	C = H = L = PF;
	Frac2 = StochK(period, 1);
	RestorePriceArrays();

	//Smoothed calculation
	PFF = AMA(Frac2, factor);
	return PFF;
}
```

**Reply by DaveSkender** [reply]
Do you know what "AMA" is here?
I'm also wondering if this is different or the same as our existing [Schaff Trend Cycle](https://daveskender.github.io/Stock.Indicators/indicators/Stc/#content) indicator.

**Comment by turok2000** [comment]
![image](https://user-images.githubusercontent.com/44271858/154012344-39475715-21e3-435a-afcb-090796efef82.png)

**Reply by DaveSkender** [reply]
Ok, thanks for the info.  I think I’ll use this as an example of how to add a custom private indicator that works identically to the ones in the library.  See #698

**Reply by turok2000** [reply]
То есть мне ждать разработки?

**Comment by turok2000** [comment]
Отличается! Мой вариант более сглажен.

**Comment by turok2000** [comment]
Прошу помочь разработать мой вариант.

**Comment by ooples** [comment]
This code is just a Schaff Trend Cycle with custom Kaufman Adaptive Moving Average as a smoother for the final result

**Reply by DaveSkender** [reply]
It's a bit odd to have `factor = 0.5` as an input value for AMA, rather than being calculated "adaptively".  Smells funny.

**Reply by turok2000** [reply]
Буду благодарен если вы покажете пример адаптивного использования factor )))

**Reply by turok2000** [reply]
Прошу помогите, запрограммировать данный код что я представил. Можно на мою почту ([email removed my moderator]) или в чате. Заранее благодарен!

**Reply by DaveSkender** [reply]
@turok2000 I'm not going to be able to write a custom indicator for you and cannot meet your needed schedule; however, I will likely add an example of how to write a custom indicator using this library (subscribe to #698 to follow) -- but again, even the example may take some time to write and share with the community.

I encourage others in this community to help out if they have time to support your unique needs.
